### DISCUSSION TITLE
Reject anonymous senders

### DISCUSSION THREAD

**Discussion by msciacca** [postalserver/postal#2345] [Help with using Postal] [answered]
Hi,

I have noticed that if someone tries to connect via Telnet to the SMTP server on port 25, they can send the email using any known registered mail on Postal without the need for any authentication whatsoever.

And postal would route that email just fine to the destination.

Isn't there a way to absolutely force authentication before being able to talk with the SMTP server, so that an anonymous connection like this isn't possible?

It seems like a security flaw where anyone could send an email on behalf of anyone.

**Comment by jamnaga** [comment]
@willpower232 can you help us?

**Comment by willpower232** [comment]
I don't think this is possible unless you have a particular incoming route which matches what people are sending as.

In the first example, I'm sending a random message to a domain Postal does not know to another domain it does not know and I am asked for authentication. In the second example, I am sending a message to a domain Postal knows is an incoming route and I am not prompted for authentication.

![image](https://user-images.githubusercontent.com/1619102/216005392-96ba7b95-3125-4b73-96c1-466d5b4214d5.png)

Are you able to tell us the commands you sent over telnet so we can see what might be happening?

**Reply by msciacca** [reply]
Hi,

In my example, I have connected via telnet as a totally anonymous user (never asked for auth). And I managed to send an email from a registered route, to a registered route.

This way, it's possible for scammers or phishing attempt to also set a name on the email so that the mail client wouldn't even show the email it's from until you click on it, and the email could look like it's totally legit to the end user. Also even after looking at where it's from it would appear as if they have written to themselves.

![Screenshot 2023-02-01 alle 10 44 23](https://user-images.githubusercontent.com/20998411/216008606-af3a73af-893a-40e1-b034-27e3a767f865.png)

**Reply by willpower232** [reply]
This makes sense though because Postal should accept all email messages for an incoming route it knows about, that is how it receives email.

You don't put a limit on your desktop email account saying who can and can't send you email messages because you won't know 100% of the addresses that legitimately need to send you email.

You can enable the spam filtering for the incoming route if you wish but I don't see what the problem is currently.

**Reply by msciacca** [reply]
Another "kinda" issue arises though.

It's possible to arbitrarily send emails between users on Postal, without the need for authentication. So if in my example this is used for a corporate email account, and it's shared between two or more companies, someone totally outside the organization could send an email as someone from another company, to someone else (basically making it look like company A is talking to company B). While in reality nobody ever sent that.

I just tried and that works, if both mail accounts are on Postal, it will allow the message through. Is that also considered normal? Shouldn't only the account holder be able to send emails from it to whatever destination he may set?

**Reply by msciacca** [reply, marked as answer]
I have made some changes to the mail_from function to serve the purpose that I described in my messages. Now it will check if the incoming email account belongs to an existing route in Postal.

If it does, it requires authentication, if authentication is provided it forwards the email over, if it can't be completed it just won't accept the email. Avoiding any possible "spoofed communication" from 3rd parties altogether. It's possible that I'm perhaps breaking RFC because I did not take the time to read it, but this does seem a more logical way to handle incoming traffic for known routes as a source mail

```
def mail_from(data)
        unless in_state(:welcomed, :mail_from_received)
          return '503 EHLO/HELO first please'
        end

        transaction_reset
        if data =~ /AUTH=/
          # Discard AUTH= parameter and anything that follows.
          # We don't need this parameter as we don't trust any client to set it
          mail_from_line = data.sub(/ *AUTH=.*/, '')
        else
          mail_from_line = data
        end

        mail_from_line_clean = mail_from_line.gsub(/MAIL FROM\s*:\s*/i, '').gsub(/.*</, '').gsub(/>.*/, '').strip
        uname, domain = mail_from_line_clean.split('@', 2)
        if uname.blank? || domain.blank?
          return '501 Invalid MAIL FROM'
        end
        uname, tag = uname.split('+', 2)
        if uname && domain && route = Route.find_by_name_and_domain(uname, domain)
          # This is a message originating from an internal route
          # We need to check that the user is allowed to send from this address
          if @credential
            @credential.use
            @state = :mail_from_received
            @mail_from = mail_from_line.gsub(/MAIL FROM\s*:\s*/i, '').gsub(/.*</, '').gsub(/>.*/, '').strip
            '250 OK'
          else
            '530 Authentication required'
          end
        else
          @state = :mail_from_received
          @mail_from = mail_from_line.gsub(/MAIL FROM\s*:\s*/i, '').gsub(/.*</, '').gsub(/>.*/, '').strip
          '250 OK'
        end
      end
```

For reference, **this was the ORIGINAL code**:
```
def mail_from(data)
        unless in_state(:welcomed, :mail_from_received)
          return '503 EHLO/HELO first please'
        end

        @state = :mail_from_received
        transaction_reset
        if data =~ /AUTH=/
          # Discard AUTH= parameter and anything that follows.
          # We don't need this parameter as we don't trust any client to set it
          mail_from_line = data.sub(/ *AUTH=.*/, '')
        else
          mail_from_line = data
        end
        @mail_from = mail_from_line.gsub(/MAIL FROM\s*:\s*/i, '').gsub(/.*</, '').gsub(/>.*/, '').strip
        '250 OK'
      end
```
